pool/paymentmgr.go

```diff
@@ -459,34 +457,140 @@ func (pm *PaymentMgr) addPaymentRequest(addr string) error {
-	// slice index when a bundle entry in the slice is removed.
-	for idx := len(bundles) - 1; idx >= 0; idx-- {
-		if bundles[idx].Total() < pm.cfg.MinPayment {
-			// Remove payments below the minimum payment if they have not been
-			// requested for by the user.
-			if !pm.isPaymentRequested(bundles[idx].Account) {
-				bundles = append(bundles[:idx], bundles[idx+1:]...)
-				continue
+	return payments, nil
+}
+
+// pendingPaymentsAtHeight fetches all pending payments at the provided height.
+func (pm *PaymentMgr) pendingPaymentsAtHeight(height uint32) ([]*Payment, error) {
+	payments := make([]*Payment, 0)
+	err := pm.cfg.DB.View(func(tx *bolt.Tx) error {
+		bkt, err := fetchPaymentBucket(tx)
+		if err != nil {
+			return err
+		}
+
+		heightBE := heightToBigEndianBytes(height)
+		paymentHeightB := make([]byte, 8)
+		cursor := bkt.Cursor()
+		for k, v := cursor.First(); k != nil; k, v = cursor.Next() {
+			_, err := hex.Decode(paymentHeightB, k[:8])
+			if err != nil {
+				return err
+			}
+
+			if bytes.Compare(heightBE, paymentHeightB) > 0 {
+				var payment Payment
+				err := json.Unmarshal(v, &payment)
+				if err != nil {
+					return err
+				}
+
+				if payment.PaidOnHeight == 0 {
+					payments = append(payments, &payment)
+				}
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return payments, nil
+}
+
+// archivedPayments fetches all archived payments. List is ordered, most
+// recent comes first.
+func (pm *PaymentMgr) archivedPayments() ([]*Payment, error) {
+	pmts := make([]*Payment, 0)
+	err := pm.cfg.DB.View(func(tx *bolt.Tx) error {
+		abkt, err := fetchPaymentArchiveBucket(tx)
+		if err != nil {
+			return err
+		}
+
+		c := abkt.Cursor()
+		for k, v := c.Last(); k != nil; k, v = c.Prev() {
+			var payment Payment
+			err := json.Unmarshal(v, &payment)
+			if err != nil {
+				return err
+			}
+			pmts = append(pmts, &payment)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return pmts, nil
+}
+
+// maturePendingPayments fetches all mature pending payments at the
+// provided height.
+func (pm *PaymentMgr) maturePendingPayments(height uint32) (map[uint32][]*Payment, error) {
+	payments := make([]*Payment, 0)
+	err := pm.cfg.DB.View(func(tx *bolt.Tx) error {
+		bkt, err := fetchPaymentBucket(tx)
+		if err != nil {
+			return err
+		}
+
+		cursor := bkt.Cursor()
+		for k, v := cursor.First(); k != nil; k, v = cursor.Next() {
+			var payment Payment
+			err := json.Unmarshal(v, &payment)
+			if err != nil {
+				return err
 			}
-			if txrules.IsDustAmount(
-				bundles[idx].Total(),
-				25, // P2PKHScriptSize
-				mempool.DefaultMinRelayTxFee) {
-				bundles = append(bundles[:idx], bundles[idx+1:]...)
+
+			spendableHeight := payment.EstimatedMaturity + 1
+			if payment.PaidOnHeight == 0 && spendableHeight <= height {
+				payments = append(payments, &payment)
 			}
 		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	pmts := make(map[uint32][]*Payment)
```

Comment:
Why are the payments being stored in a map like this? I don't see the need for using a map with the height as the key. - the key doesnt seem to used anywhere other than tests.

Reply:
The purpose of that was to group payments by the height they were created on, I'll be reworking that actually. needs to be keyed by the coinbase they are paying from. That has more utility.

Reply:
used the block hash as the key to make it easier to prune payments sourcing from orphaned blocks.